src/contracts/examples/intents-example/SwapIntentInvertBidDAppControl.sol

```diff
@@ -0,0 +1,180 @@
+        require(msg.sender == ATLAS, "SwapIntentDAppControl: InvalidSender");
+        require(_addressPointer() == CONTROL, "SwapIntentDAppControl: InvalidLockState");
+        require(address(this) != CONTROL, "SwapIntentDAppControl: MustBeDelegated");
+
+        address user = _user();
+
+        require(
+            _availableFundsERC20(
+                swapIntent.tokenUserSells, user, swapIntent.maxAmountUserSells, ExecutionPhase.PreSolver
+            ),
+            "SwapIntentDAppControl: SellFundsUnavailable"
+        );
+
+        return SwapIntent({
+            tokenUserBuys: swapIntent.tokenUserBuys,
+            tokenUserSells: swapIntent.tokenUserSells,
+            amountUserBuys: swapIntent.amountUserBuys,
+            maxAmountUserSells: swapIntent.maxAmountUserSells
+        });
+    }
+
+    // ---------------------------------------------------- //
+    //                     Atlas hooks                      //
+    // ---------------------------------------------------- //
+
+    /*
+    * @notice This function is called before a solver operation executes
+    * @dev This function is delegatecalled: msg.sender = Atlas, address(this) = ExecutionEnvironment
+    * @dev It transfers the tokens that the user is selling to the solver
+    * @param solverOp The SolverOperation that is about to execute
+    * @return true if the transfer was successful, false otherwise
+    */
+    function _preSolverCall(SolverOperation calldata solverOp, bytes calldata returnData) internal override {
+        address solverTo = solverOp.solver;
+        if (solverTo == address(this) || solverTo == _control() || solverTo == ATLAS) {
+            revert();
+        }
+
+        SwapIntent memory swapData = abi.decode(returnData, (SwapIntent));
+
+        // The solver must be bidding less than the intent's maxAmountUserSells
+        require(solverOp.bidAmount <= swapData.maxAmountUserSells, "SwapIntentInvertBid: BidTooHigh");
+
+        // Optimistically transfer to the solver contract the amount that the solver is invert bidding
+        _transferUserERC20(swapData.tokenUserSells, solverTo, solverOp.bidAmount);
+
+        return; // success
+    }
+
+    /*
+    * @notice This function is called after a solver operation executed
+    * @dev This function is delegatecalled: msg.sender = Atlas, address(this) = ExecutionEnvironment
+    * @dev It transfers to the user the tokens they are buying
+    * @param _
+    * @param returnData The return data from the user operation (swap data)
+    * @return true if the transfer was successful, false otherwise
+    */
+    function _postSolverCall(SolverOperation calldata, bytes calldata returnData) internal override {
+        SwapIntent memory swapIntent = abi.decode(returnData, (SwapIntent));
+        uint256 buyTokenBalance = ERC20(swapIntent.tokenUserBuys).balanceOf(address(this));
+
+        if (buyTokenBalance < swapIntent.amountUserBuys) {
+            revert();
```

Comment:
Can we add a revert message here (and standardize on one type of error style) and make sure we trigger it in tests if solver doesn't transfer back the bought tokens